src/lm_robust_helper.cpp

```diff
@@ -426,16 +428,30 @@ List lm_variance(Eigen::Map<Eigen::MatrixXd>& X,
               H2s.block(0, p_pos, r_fe, r_fe) = ME * X.block(start_pos, 0, len, r_fe) * M_U_ct;
               H3s.block(0, p_pos, r_fe, r_fe) = MEU * Omega_ct;
             }
+
+          } else if (se_type == "CR3") {
+
+            Eigen::CompleteOrthogonalDecomposition<Eigen::MatrixXd> At(
+                Eigen::MatrixXd::Identity(len, len) -
+                  Xoriginal.block(start_pos, 0, len, r_fe) *
+                  meatXtX_inv *
+                  X.block(start_pos, 0, len, r_fe).transpose()
+            );
+
+            if (Xunweighted.isNotNull()) {
+              At_WX_inv =  At.solve(Eigen::MatrixXd::Identity(len, len)).transpose() * X.block(start_pos, 0, len, r_fe);
```

Comment:
Why solving against Identity? Maybe add a comment.

Reply:
In both case we need the transpose, but unweighted case is symmetric so it doesn't matter. I could instead transpose the matrix before hand (as you can't transpose the `CompleteOrthogonalDecomposition`).

Reply:
If Eigen knows the solution is sym, the transpose should be a no-op - I generally prefer the more concise and general expression over the two branches that are slightly different. However, there's a numerical and performance penalty to creating the inverse explicitly when it isn't actually needed. 


Maybe you can move the transpose inside a la https://math.stackexchange.com/a/340234/139343

Reply:
It doesn't know the solution is symmetric, so the transpose is not a no-op, right?

Reply:
It /should/ know if an object is symmetric or not. When inner products are used to calculate eg covariance matrices, the result is  supposed to be subclassed as self-adjoint and further operators dispatched to symmetric implementations. It's an easy 50% speed up for many problems (eg crossprod(X) vs t(X) %*% X, chol(X) vs qr(X), etc). 

Even if you store a symmetric matrix in a regular Matrix variable, it's methods will still be dispatched to symmetric functions appropriately, which can then also return symmetric matrices will their own appropriate methods (this is `covariance` in the CS parlance https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science) ).

See the RcppEigen Vignette section on Crossproduct of a matrix -  IIRC that's roughly where you had started in the first place, but I might be thinking of Armadillo. There's also a chance I'm getting the two libraries confused.

Reply:
also transposes are basically free compared to everything else you're doing.